### PR TITLE
[stdlib]Un-revert string comparison

### PR DESCRIPTION
**Pull request by lancep**
This brings back the string comparison change along with a fix for opaque substrings.

### CODE REVIEW

**Comment by lancep** [comment]
Please test with the following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1441

@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please test

**Comment by lancep** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please clean test Linux platform

**Comment by lancep** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please test Linux platform

**Comment by lancep** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please test

**Comment by eeckstein** [comment]
@lancep Does this fix the sporadic failures which we saw on the bots?

File: stdlib/public/stubs/UnicodeNormalization.cpp [hunk 1]
```diff
@@ -288,8 +288,8 @@ swift::__swift_stdlib_unorm2_getNFCInstance(__swift_stdlib_UErrorCode *err) {
 }
 
 int32_t swift::__swift_stdlib_unorm2_normalize(
-    const __swift_stdlib_UNormalizer2 *norm, const __swift_stdlib_UChar *src,
-    __swift_int32_t len, __swift_stdlib_UChar *dst, __swift_int32_t capacity,
+    const __swift_stdlib_UNormalizer2 *norm, const __swift_int32_t *src,
```
Comment (milseman) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: You mean `uint16_t`? Why are we doing all this churn anyways?

Comment (lancep) [reply stdlib/public/stubs/UnicodeNormalization.cpp]: Good catch. I'm not sure why we started #if-ing the typedef for __swift_stdlib_UChar. Maybe the ICU headers differ on some other flavor of Linux?

**Comment by lancep** [comment]
@eeckstein I don't think those were related to this change now. I'm still investigating why occasionally String's capacity is higher than we expect when appending.

**Comment by lancep** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please test

**Comment by lancep** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please test

**Comment by lancep** [comment]
@eeckstein This just hit that string capacity test failure again. It may be a coincidence, but it's also possible that these unrelated changes affected the test somehow. I've seen this in the past with the Array tests. It seems that on Linux, malloc occasionally will give us more space when resizing the array than it does on Darwin platforms. I'll investigate more on Monday, but I've change the test to get it passing for now, by bumping it from expecting to be <= 40 instead of 34.

**Comment by lancep** [comment]
Please test with the following pull request:
apple/swift-corelibs-foundation#1441

@swift-ci please test

**Comment by jrose-apple** [comment]
We're seeing [failures](https://ci.swift.org/view/swift-5.0-branch/job/oss-swift-5.0-incremental-RA-linux-ubuntu-16_10/239/) from this on Linux on the 5.0 branch, since corelibs Foundation doesn't have a separate 5.0 branch at the moment. What should we do about this? cc @shahmishal, @bob-wilson, @parkera

**Comment by shahmishal** [comment]
@bob-wilson Should we just have @lancep cherry-pick the change from swift:master into swift:swift-5.0-branch to unblock testing?

**Comment by lancep** [comment]
We would need to merge @milseman's StringGuts changes there as well.

**Comment by shahmishal** [comment]
@lancep, Can you cherry-pick changes to swift-5.0-branch?

**Comment by lancep** [comment]
I've opened https://github.com/apple/swift-corelibs-foundation/pull/1445 to skip the test in question

**Pull request merged**